### ISSUE TITLE
`torch.profiler.tensorboard_trace_handler` does not save trace with `torch.profiler.schedule`

### ISSUE THREAD

**Issue by Zymrael** [pytorch/pytorch#91682]
### 🐛 Describe the bug

This does not save a `.json` trace:

```
import torch 
import torch.nn as nn

f = nn.Linear(1, 1)
x = torch.ones(1, 1)

with torch.profiler.profile(
    activities=[torch.profiler.ProfilerActivity.CPU, torch.profiler.ProfilerActivity.CUDA],
    record_shapes=True,
    with_stack=True,
    profile_memory=True, 
    schedule=torch.profiler.schedule(wait=1, warmup=1, active=3, repeat=2),
    on_trace_ready=torch.profiler.tensorboard_trace_handler('.', worker_name=None),
    ) as prof:
    f(x)
 ```
 
 The same code without schedule works flawlessly:
 
 ```
 import torch 
import torch.nn as nn

f = nn.Linear(1, 1)
x = torch.ones(1, 1)

with torch.profiler.profile(
    activities=[torch.profiler.ProfilerActivity.CPU, torch.profiler.ProfilerActivity.CUDA],
    record_shapes=True,
    with_stack=True,
    profile_memory=True, 
    on_trace_ready=torch.profiler.tensorboard_trace_handler('.', worker_name=None),
    ) as prof:
    f(x)
 ```
 
 From the [docs](https://pytorch.org/docs/stable/profiler.html#torch.profiler.tensorboard_trace_handler) it seems both should work.

### Versions

```
Collecting environment information...
PyTorch version: 1.13.0a0+d0d6b1f
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.24.3
Libc version: glibc-2.31

Python version: 3.8.13 | packaged by conda-forge | (default, Mar 25 2022, 06:04:10)  [GCC 10.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-1082-kvm-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100 80GB PCIe
GPU 1: NVIDIA A100 80GB PCIe
GPU 2: NVIDIA A100 80GB PCIe
GPU 3: NVIDIA A100 80GB PCIe
GPU 4: NVIDIA A100 80GB PCIe
GPU 5: NVIDIA A100 80GB PCIe
GPU 6: NVIDIA A100 80GB PCIe
GPU 7: NVIDIA A100 80GB PCIe

Nvidia driver version: 510.108.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.6.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] functorch==0.3.0a0
[pip3] numpy==1.22.2
[pip3] pytorch-fast-transformers==0.4.0
[pip3] pytorch-lightning==1.7.7
[pip3] pytorch-quantization==2.1.2
[pip3] torch==1.13.0a0+d0d6b1f
[pip3] torch-tensorrt==1.2.0a0
[pip3] torchmetrics==0.9.3
[pip3] torchtext==0.11.0a0
[pip3] torchvision==0.14.0a0
[conda] functorch                 0.3.0a0                  pypi_0    pypi
[conda] mkl                       2020.4             h726a3e6_304    conda-forge
[conda] mkl-include               2020.4             h726a3e6_304    conda-forge
[conda] numpy                     1.22.2           py38h6ae9a64_0    conda-forge
[conda] pytorch-fast-transformers 0.4.0                    pypi_0    pypi
[conda] pytorch-lightning         1.7.7                    pypi_0    pypi
[conda] pytorch-quantization      2.1.2                    pypi_0    pypi
[conda] torch                     1.13.0a0+d0d6b1f          pypi_0    pypi
[conda] torch-tensorrt            1.2.0a0                  pypi_0    pypi
[conda] torchmetrics              0.9.3                    pypi_0    pypi
[conda] torchtext                 0.11.0a0                 pypi_0    pypi
[conda] torchvision               0.14.0a0                 pypi_0    pypi
```

cc @robieta @chaekit @aaronenyeshi @ngimel @nbcsm @guotuofeng @guyang3532 @gaoteng-git @tiffzhaofb

**Comment by Superkeyv** [comment]
you can try add `prof.step()` to every step

```
import torch 
import torch.nn as nn

f = nn.Linear(1, 1)
x = torch.ones(1, 1)

with torch.profiler.profile(
    activities=[torch.profiler.ProfilerActivity.CPU, torch.profiler.ProfilerActivity.CUDA],
    record_shapes=True,
    with_stack=True,
    profile_memory=True, 
    schedule=torch.profiler.schedule(wait=1, warmup=1, active=3, repeat=2),
    on_trace_ready=torch.profiler.tensorboard_trace_handler('.', worker_name=None),
    ) as prof:
    f(x)
    # add this
    prof.step()
```

**Comment by Zymrael** [comment]
I had tried this - does not work for my env.

**Comment by Superkeyv** [comment]
ok, the real problem is

```
schedule=torch.profiler.schedule(wait=1, warmup=1, active=3, repeat=2),
```

which means, cycle of [wait, warmup,active] has repeat times. e.g
wait step 0, warmup step 1, and record profile at step 2, 3, 4
then repeat cycle,
wait step 5, warmup step 6, and record profile at step 7, 8, 9


so, at your code, there's only step 0.

I can show you a full version like:

```
import torch 
import torch.nn as nn

f = nn.Linear(1, 1)
x = torch.ones(1, 1)

with torch.profiler.profile(
    activities=[torch.profiler.ProfilerActivity.CPU, torch.profiler.ProfilerActivity.CUDA],
    record_shapes=True,
    with_stack=True,
    profile_memory=True, 
    schedule=torch.profiler.schedule(wait=1, warmup=1, active=3, repeat=2),
    on_trace_ready=torch.profiler.tensorboard_trace_handler('.', worker_name=None),
    ) as prof:
    for i in range(max_steps):
      f(x)
      prof.step()
```

**Comment by Zymrael** [comment]
Of course, forgot to remove wait and warmup cycles. Thanks!
